engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java

```diff
@@ -3808,7 +3808,7 @@ public boolean upgradeVmDb(final long vmId, final ServiceOffering newServiceOffe
         if (currentServiceOffering.isDynamic() && !newServiceOffering.isDynamic()) {
             removeCustomOfferingDetails(vmId);
         }
-        VMTemplateVO template = _templateDao.findById(vmForUpdate.getTemplateId());
+        VMTemplateVO template = _templateDao.findByIdIncludingRemoved(vmForUpdate.getTemplateId());
```

Comment:
does a VM exists after its template got deleted?

Reply:
-1, this may not be the right fix, it checks to update a VM whose template is deleted.

Reply:
> does a VM exists after its template got deleted?

@sureshanaparti 
yes, template can be forced deleted (from secondary storage), but the base image of template still exist on primary storage, vms are not impacted (reinstall might not work)

Reply:
got it @weizhouapache , but is it right thing to consider the dynamic scaling setting from the deleted template?

Reply:
@sureshanaparti
I think it should be ok. If template is configured to support dynamic scaling, the VM image should support it as well because the vm image is created from the template.

Reply:
yes, agree with @weizhouapache

Reply:
@sureshanaparti would you like to review this again

Reply:
@weizhouapache @harikrishna-patnala If VM existence (and it's scaling) is allowed after template deletion, better to skip dynamically scalable flag in that template (and consider as 'true'). If one wants to scale the VM when the flag is false, it is not possible to update the flag on the deleted template. What's your thoughts?

Reply:
@sureshanaparti, as @shwstppr mentioned we use deleted template properties in other places also. So this procedure looks good to me and we cannot consider true because it can happen that the deleted template was not dynamically scalable. So better use the template properties as it is

Reply:
lgtm, discussed offline with @harikrishna-patnala